Fix `take` and `drop` on very large `NumericRange`s

AminMal opened the pull request:
This PR aims to prevent the evaluation of `NumericRange$length` to be able to perform `take` and `drop` on very large ranges.

Fixes scala/bug#12706

SethTisue commented:
sequel to #10251

Ichoran commented on src/library/scala/collection/immutable/NumericRange.scala:
```diff
@@ -140,14 +140,16 @@ sealed class NumericRange[T](
     if (n <= 0 || isEmpty) newEmptyRange(start)
-    else if (n >= length) this
-    else new NumericRange.Inclusive(start, locationAfterN(n - 1), step)
+    else if (num.lt(locationAfterN(n - 1), end))
```
This will overflow in case of `Int` or `Long`.

For instance, consider `NumericRange(2000000000, 2100000000, 70000000).take(20)`.

Ichoran commented:
I appreciate your motivation to supply a fix, and it will be great if you can make one, but honestly, you're going to have to think much more carefully about numeric overflow if you're going to succeed.

When you don't know if your calculations will fit, you have to think about _every single operation_, even a +1, and ask: can this overflow?  What happens if it does?

The answers are usually "yes" and "the result is badly wrong", respectively.  That's what makes this code so tricky (it has a long history of being broken in various ways, almost all due to overflow issues).

Ichoran commented:
Here is a partial list of things that you cannot count on, given an integral type `T`.

* `x` and `-x` can both be represented.  (Example: `-Int.MinValue` overflows and is `Int.MinValue` again.)
* The length between to items of type `T` can be represented with `T`.  (Example: `Int.MinValue` to `Int.MaxValue` with step size 1)
* The difference between endpoints of type `T` can be represented with `T`. (Example: `Int.MaxValue - Int.MinValue` does not fit in an `Int`.)
* Multiplication of a step size by anything except 0 or 1 will fit in `T`.  (Example: with step size `2000000000`, only multiplication by -1, 0, or 1 fits in `Int`; everything else overflows.)

It's very finicky.

If you can understand why each piece of `count` is the way it is--it might not be the best possible, but everything there has a reason, and most of the reasons are documented in comments--then you're on the right track.

AminMal commented:
> Here is a partial list of things that you cannot count on, given an integral type `T`.
> 
> * `x` and `-x` can both be represented.  (Example: `-Int.MinValue` overflows and is `Int.MinValue` again.)
> * The length between two items of type `T` can be represented with `T`.  (Example: `Int.MinValue` to `Int.MaxValue` with step size 1)
> * The difference between endpoints of type `T` can be represented with `T`. (Example: `Int.MaxValue - Int.MinValue` does not fit in an `Int`.)
> * Multiplication of a step size by anything except 0 or 1 will fit in `T`.  (Example: with step size `2000000000`, only multiplication by -1, 0, or 1 fits in `Int`; everything else overflows.)
> 
> It's very finicky.
> 
> If you can understand why each piece of `count` is the way it is--it might not be the best possible, but everything there has a reason, and most of the reasons are documented in comments--then you're on the right track.

Thank you for reminding me of the edge cases. I know that every piece of code in `count` is meant for something, and I don't want to change anything in it :). It just doesn't make sense to use it everywhere, right? why would we need the count of the current range when it's already constructed and we want to take or drop some members from it?

I'm coming up with another commit which fixes the edge cases, with more unit tests and property tests.

Ichoran commented:
You don't necessarily need the count itself, but if you take more than the count, the range should be unchanged, and if you drop more than the count, the range should be empty.  Getting that to work without actually knowing the count, or at least doing most of the same calculations, seems difficult.

AminMal commented:
@Ichoran @lrytz  I came up with an idea to split a wide range into two ranges, in which the subtraction and addition in `T` would be safe, and it works fine with all the unit tests and the new pbt for `take` and `drop`. Do you think that this would also be useful for `NumericRange.count` as well?

For instance, `[Int.MinValue, Int.MaxValue)` is split into `left = [Int.MinValue, -1]` and `right = [0, Int.MaxValue)`, and then using that simple implementation for the length of the range in my first PR (with some improvements) would be safe.

Ichoran commented on src/library/scala/collection/immutable/NumericRange.scala:
```diff
@@ -132,6 +132,59 @@ sealed class NumericRange[T](
+    val sameSign = num.equiv(num.sign(start), num.sign(end))
+
+    if (eitherOfBoundariesIsZero || sameSign) { // subtraction is safe
+      val len = unsafeRangeLength(this)
```
This will fail on `Int.MinValue` to `0`.

You said that you were going to split the ranges into MinValue, -1 and 0, MaxValue.  But here you're testing for zero, not -1.  Instead, you might want to test both endpoints for `num.lt(_, num.zero)`

AminMal replied:
Great point! I'll fix it right now! I think testing both start and end with `num.lt(_, zero)` is redundant, I should've never used that actually. I mean, it's already handled in the else branch where it splits the range into two sub-ranges, this way it has less branches/conditions so I think it's more readable, and there's actually no need to check them.

Ichoran commented on src/library/scala/collection/immutable/NumericRange.scala:
```diff
@@ -132,6 +132,59 @@ sealed class NumericRange[T](
+    if (num.equiv(start, end) || n == 0) return n >= 1
+
+    val smallerThanInt = num match {
+      case _: Numeric.ShortIsIntegral | _: Numeric.ByteIsIntegral => true
```
Couldn't you just do the whole calculation here, exactly, with `Int`s, and be done with that part rather than coming back to it later?

AminMal replied:
I actually wrote it this way at first, then I realized having a general algorithm/solution would be more reliable and readable rather than writing different logics for different generic types. We just leave the final calculations type-specific, and the logic stays the same for every type `T`. Also, that way I had to implement two different `unsafeLength` methods, one which works for small types and returns `Int`, another one which returns `T`. Let me know if you think the other way.

Ichoran replied:
The unity of having a single algorithm sounds like a benefit.  Makes sense to keep it this way, then!

AminMal replied:
Alright, resolving this as well!

Ichoran commented on src/library/scala/collection/immutable/NumericRange.scala:
```diff
@@ -132,6 +132,59 @@ sealed class NumericRange[T](
+      // split to two ranges, which subtraction is safe in both of them (around zero)
+      val closestToZeroFromStart = num.rem(start, step)
+
+      val (l: NumericRange[T], r: NumericRange[T]) = if (num.lt(num.sign(start), num.zero)) {
```
The call to `sign` here is superfluous.

Ichoran commented on src/library/scala/collection/immutable/NumericRange.scala:
```diff
@@ -132,6 +132,59 @@ sealed class NumericRange[T](
+      if (smallerThanInt) n >= num.toInt(len) else num.gteq(num.fromInt(n), len)
+    } else {
+      // split to two ranges, which subtraction is safe in both of them (around zero)
+      val closestToZeroFromStart = num.rem(start, step)
```
This gives an unsafe range if start = `Int.MinValue` and `step` is a power of two.

AminMal replied:
No it's fine, for instance taking 2 as `step`, it takes you `2**31` steps to make it exactly to zero. if it's 4, it takes `2**30` steps and so on.

Ichoran commented on src/library/scala/collection/immutable/NumericRange.scala:
```diff
@@ -132,6 +132,59 @@ sealed class NumericRange[T](
+
+      val (l: NumericRange[T], r: NumericRange[T]) = if (num.lt(num.sign(start), num.zero)) {
+        val lastElementBeforeZero =
+          if (num.equiv(closestToZeroFromStart, num.zero)) step else closestToZeroFromStart
```
This seems to assume that the sign of `start` and `step` is the same.  But we don't know what the signs are.

AminMal replied:
Fixed this as well!

Ichoran commented on src/library/scala/collection/immutable/NumericRange.scala:
```diff
@@ -132,6 +132,59 @@ sealed class NumericRange[T](
+        val lastElementBeforeZero =
+          if (num.equiv(closestToZeroFromStart, num.zero)) step else closestToZeroFromStart
+
+        (NumericRange.inclusive(start, zero - num.abs(lastElementBeforeZero), step), copy(zero, end, step))
```
Is this correct if `start` is, say, `Int.MaxValue`?

AminMal replied:
It's in the branch where start is negative

Ichoran reviewed (changes requested):
This is much better!  But there are still a bunch of edge cases having to do with extreme values for start/end and probably also with inverted ranges (start > end, step < 0).  I didn't find all of them, but I've at least given a flavor of the remaining concerns.

Ichoran commented:
In principle, code like this could replace `count`, once the edge cases are all worked out.

In practice, I wouldn't recommend it, for two reasons.  (1) `count` is battle-tested and apparently works; even if all the tests pass, this won't have been tested in the wild.  (2) This code isn't that much shorter than `count` anyway.

AminMal commented:
> In principle, code like this could replace `count`, once the edge cases are all worked out.
> 
> In practice, I wouldn't recommend it, for two reasons. (1) `count` is battle-tested and apparently works; even if all the tests pass, this won't have been tested in the wild. (2) This code isn't that much shorter than `count` anyway.

@Ichoran What I meant is it can be broken down into smaller functions I believe. There's a huge amount of the code in `count` which just tries to see if the length is more than `Int.MaxValue`, and it doesn't always work correctly for all generic `T` types, and personally, I don't see it much readable. Like what we're having a conversation for at scala/bug#12716. It can easily just call `crossesTheEndAfter`. Imagine reading the `count`, and see this small piece of code:
```scala
if (crossesTheEndAfter(Int.MaxValue)) throw whatever else calculateTheRangeSafely
```
You just know what this block/expression does right away! You shouldn't be reading many lines of code just to understand that we're going to check the integer limit overflow.

AminMal commented:
> You might have a span of longer than `T` can represent, which crosses zero, but has fewer than `Int.MaxValue` items. So knowing that you will fall off within `Int.MaxValue` steps still won't tell you how to actually calculate the length!

Of course, I'm just talking about the idea, imagine something like this:
```scala
def getNumericRangeLeftAndRightLengths[T](r: NumericRange[T]): (T, T) = { // left and right ranges length }

// somewhere in count
val (leftLength, rightLength) = getLeftAndRightLengthsForRange(r)

if (num.gteq(num.minus(leftLength, num.fromInt(Int.MaxValue), rightLength))) 
  throw ...
else num.toInt(num.plus(leftLength, rightLength))
```
Yeah of course it requires more precise considerations, or some modifications or something, but I think at the end it would be nice to be able to break down the count.

> By the way, you said that you had the idea of dividing the range into two pieces

Yeah I hadn't fully read the what count does, and I didn't claim that I invented such thing :) that was the idea I came up with in the new commit.

Ichoran commented on src/library/scala/collection/immutable/NumericRange.scala:
```diff
@@ -132,6 +132,60 @@ sealed class NumericRange[T](
+      if (!isInclusive && num.equiv(remainder, num.zero)) quotient else num.plus(quotient, num.one)
+    }
+    val stepIsInTheSameDirectionAsStartToEndVector =
+      (num.gt(end, start) && num.sign(step) == num.one) || (num.lt(end, start) && num.sign(step) == -num.one)
```
It's probably more efficient to use `num.gt(step, num.zero)` than `num.sign(step) == num.one` because the former creates a new (boxed) `T`.  Likewise with `lt`, but moreso, because `-num.one` also creates a new boxed `T`.  (In general.  Specific `T`s may cache small values.)

Ichoran commented on src/library/scala/collection/immutable/NumericRange.scala:
```diff
@@ -132,6 +132,60 @@ sealed class NumericRange[T](
+    if (num.equiv(start, end) || n <= 0 || !stepIsInTheSameDirectionAsStartToEndVector) return n >= 1
+
+    val smallerThanInt = num match {
+      case _: Numeric.ShortIsIntegral | _: Numeric.ByteIsIntegral => true
```
This won't work for small custom types; this is why `count` does it a different way (actually testing whether `Int.MaxValue` survives a bidirectional trip).  If the algorithm works regardless, maybe just drop this case entirely?

Ichoran replied:
It doesn't look like it will work regardless, but maybe defer this comparison to the end?

AminMal replied:
Hmm...Yeah makes sense. I can think of a function which detects whether a `T` fits in `Int`s, until this point we know the length in `T`, so the function can do the trick. (Bases on your hint on `NumericRange.count`)

Ichoran commented on src/library/scala/collection/immutable/NumericRange.scala:
```diff
@@ -132,6 +132,60 @@ sealed class NumericRange[T](
+        (NumericRange.inclusive(start, zero - num.abs(lastElementBeforeZero), step), copy(zero, end, step))
+      } else {
+        val firstElementAfterZero =
+          if (num.equiv(closestToZeroFromStart, num.zero)) num.abs(step) else num.abs(closestToZeroFromStart)
```
Taking the absolute value of a remainder is basically never the right idea.  the remainder of -7 when divided by 3 might be -1 or it might be 2 (depending on whether you insist remainders have to be positive), but never 1.  Likewise, the remainder of 8 divided by 3 might be 2 or it might be -1, but certainly not 1.

So `num.abs(closestToZeroFromStart)` isn't the right thing to do.

I _think_ we can count on the sign of the remainder matching the sign of the number being divided, in which case `num.abs` can be dropped, but if we _can't_, the `abs` isn't going to fix it; instead, it would give the wrong step size when crossing 0.

AminMal replied:
There were 2 or 3 completely redundant function calls in this part of the code, just removed them.

Ichoran commented on src/library/scala/collection/immutable/NumericRange.scala:
```diff
@@ -132,6 +132,60 @@ sealed class NumericRange[T](
+        n >= num.toInt(leftLength) + num.toInt(rightLength)
+      } else {
+        // instead of n >= rightLength + leftLength which may cause addition overflow,
+        // this can be used (n - rightLength) >= leftLength
```
The right/left on the comment doesn't match the right/left in the code.  Either way works, but better not to say one and do the other!

Ichoran reviewed (changes requested):
This is looking very close to something that ought to work robustly!

Ichoran commented:
@AminMal - Thanks for having the patience to stick with this surprisingly difficult part of the code!

AminMal commented:
@Ichoran Not sure if we discussed this earlier, but seems like there are some issues when splitting ranges that start from `T`'s MinValue, or end with MaxValue. Since I split them into almost something like this: `[Min, -1] ++ [0, Max]`. This is also fixed in my new commit coming up in a few minutes, in addition to your latest change requests.

Ichoran commented on src/library/scala/collection/immutable/NumericRange.scala:
```diff
@@ -132,6 +132,75 @@ sealed class NumericRange[T](
+      so we take two steps smaller than zero to ensure unsafeRangeLength won't overflow (taking one step may overflow depending on the step).
+      Same thing happens for MaxValue from zero, so we take one step further to ensure the safety of unsafeRangeLength.
+      After performing such operation, there are some elements remaining in between and around zero,
+      which their length is represented by curry.
```
Why is this called curry?  Do you mean carry?

AminMal replied:
Oh yeah that's so silly :). I'll make sure to fix it, and write those tests for `BigInt`.

Ichoran reviewed (approved):
I can't spot any other remaining issues just by reading the code (in the amount of time I'm willing to spend staring at code).  I recommend testing this with a custom `NumericRange` (e.g. backed by `BigInt` but limited to some smaller range) if you have the time to create one, to catch any additional weird behavior that might not be revealed by testing `Long` and `Short` and so on.  But regardless, it looks like a clear improvement on the current state of affairs, and seems unlikely to deliver any easily-triggered bugs.

Good job!

Ichoran commented:
I'm paying less attention to this issue now since it seems pretty much solved, so you might need to ping me on another channel if it needs my attention again.

SethTisue commented:
@AminMal This appears ready for merge. Before we do that, mind squashing this into fewer commits? (Unless you feel that the present sequence of commits represents distinct changes that are easier to understand as a series.)

som-snytt commented:
Recommend squash to one commit, keeping previous messages in the new message body. (The project encourages commit message bodies.) Esp as the additions are green blocks and not scattered. 
![image](https://user-images.githubusercontent.com/369425/219550048-f03fe336-b105-4086-8a30-4fce5bc7f8c8.png)

AminMal committed:
Implemented crossesTheEndAfterN, and used it in take and drop

SethTisue commented:
thank you @AminMal and @Ichoran for your amazing patience and attention to detail on this one

lrytz commented:
Got a test failure in a recent merge commit:

https://github.com/scala/scala/commit/edfb5646f23fc802b315540b41f52e17b8d3d663

https://github.com/scala/scala/actions/runs/4233052929/jobs/7353522195#step:6:7331

```
[error] Test scala.collection.immutable.NumericRangeTest.numericRangeSmallTypesTake failed: java.lang.AssertionError: null, took 0.001 sec
[error]     at scala.collection.immutable.NumericRangeTest.numericRangeSmallTypesTake(NumericRangeTest.scala:230)
[error]     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[error]     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[error]     at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[error]     at java.lang.reflect.Method.invoke(Method.java:566)
[error]     ...
```

AminMal commented:
> Got a test failure in a recent merge commit:
> 
> [edfb564](https://github.com/scala/scala/commit/edfb5646f23fc802b315540b41f52e17b8d3d663)
> 
> https://github.com/scala/scala/actions/runs/4233052929/jobs/7353522195#step:6:7331
> 
> ```
> [error] Test scala.collection.immutable.NumericRangeTest.numericRangeSmallTypesTake failed: java.lang.AssertionError: null, took 0.001 sec
> [error]     at scala.collection.immutable.NumericRangeTest.numericRangeSmallTypesTake(NumericRangeTest.scala:230)
> [error]     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> [error]     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> [error]     at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> [error]     at java.lang.reflect.Method.invoke(Method.java:566)
> [error]     ...
> ```

Hmmm, the fact that it succeeds on other versions of Ubuntu, Java and even Windows is kind of strange. The only thing that I can think of is the random size selection part (which now that I think, actually makes not much sense in unit tests). I can change that part of the test, but how can I assure that it would work on that specific JVM and OS? And would I need to create another PR? @SethTisue

SethTisue commented:
You'd need to create a new PR to make any further changes, yes.

I believe it's highly likely that the failure wasn't related to the OS version or Java version, but was random. I ran the test about 20 or 30 times locally and it didn't fail, but you might try running it in a loop and letting it go for a while.

AminMal commented:
> You'd need to create a new PR to make any further changes, yes.
> 
> I believe it's highly likely that the failure wasn't related to the OS version or Java version, but was random. I ran the test about 20 or 30 times locally and it didn't fail, but you might try running it in a loop and letting it go for a while.

I did this `for (take <- 0 to Byte.MaxValue) { testcase }`, and turns out that it fails for 0! It was so silly, I'll create another PR right now and remove the random values.

SethTisue commented:
Cool, thanks!

We tend to make JUnit tests (in the `test/junit` directory) deterministic, and use ScalaCheck (in the `test/scalacheck` directory) when we want random input. Both approaches have a place.

SethTisue commented:
followup PR is #10322
